internal/oci/runtime_oci.go

```diff
@@ -898,73 +899,89 @@ func (r *runtimeOCI) AttachContainer(c *Container, inputStream io.Reader, output
+			return errors.Wrapf(err, "dialing %d", port)
 		}
-	}()
-	containerPid := c.State().Pid
-	socatPath, lookupErr := exec.LookPath("socat")
-	if lookupErr != nil {
-		return fmt.Errorf("unable to do port forwarding: socat not found")
-	}
+		defer conn.Close()
 
-	args := []string{"-t", fmt.Sprintf("%d", containerPid), "-n", socatPath, "-", fmt.Sprintf("TCP4:localhost:%d", port)}
+		errCh := make(chan error, 2)
 
-	nsenterPath, lookupErr := exec.LookPath("nsenter")
-	if lookupErr != nil {
-		return fmt.Errorf("unable to do port forwarding: nsenter not found")
-	}
+		debug := func(format string, args ...interface{}) {
+			log.Warnf(ctx, fmt.Sprintf(
+				"PortForward (id: %s, port: %d): %s", c.ID(), port, format,
+			), args...)
+		}
 
-	commandString := fmt.Sprintf("%s %s", nsenterPath, strings.Join(args, " "))
-	logrus.Debugf("executing port forwarding command: %s", commandString)
+		// Copy from the the namespace port connection to the client stream
+		go func() {
+			debug("copy data from container to client")
+			_, err := io.Copy(stream, conn)
+			errCh <- err
+		}()
 
-	command := exec.Command(nsenterPath, args...)
-	if v, found := os.LookupEnv("XDG_RUNTIME_DIR"); found {
-		command.Env = append(command.Env, fmt.Sprintf("XDG_RUNTIME_DIR=%s", v))
-	}
-	command.Stdout = stream
-
-	stderr := new(bytes.Buffer)
-	command.Stderr = stderr
-
-	// If we use Stdin, command.Run() won't return until the goroutine that's copying
-	// from stream finishes. Unfortunately, if you have a client like telnet connected
-	// via port forwarding, as long as the user's telnet client is connected to the user's
-	// local listener that port forwarding sets up, the telnet session never exits. This
-	// means that even if socat has finished running, command.Run() won't ever return
-	// (because the client still has the connection and stream open).
-	//
-	// The work around is to use StdinPipe(), as Wait() (called by Run()) closes the pipe
-	// when the command (socat) exits.
-	inPipe, err := command.StdinPipe()
-	if err != nil {
-		return fmt.Errorf("unable to do port forwarding: error creating stdin pipe: %v", err)
-	}
-	var copyError error
-	go func() {
-		emptyStreamOnError = false
-		_, copyError = pools.Copy(inPipe, stream)
-		inPipe.Close()
-	}()
+		// Copy from the client stream to the namespace port connection
+		go func() {
+			debug("copy data from client to container")
+			_, err := io.Copy(conn, stream)
+			errCh <- err
+		}()
+
+		// Wait until the first error is returned by one of the connections we
+		// use errFwd to store the result of the port forwarding operation if
+		// the context is cancelled close everything and return
+		var errFwd error
+		select {
+		case errFwd = <-errCh:
+			debug("stop forwarding in direction: %v", errFwd)
+		case <-ctx.Done():
+			debug("cancelled: %v", ctx.Err())
+			return ctx.Err()
+		}
 
-	runErr := command.Run()
+		// give a chance to terminate gracefully or timeout
+		const timeout = 1 * time.Second
```

Comment:
Increasing the timeout to 1s fixes the failing e2e tests at least on my local machine.

Reply:
great catch Sascha, I applied the same timeout to containerd and it fixes the e2e test.
Please, can you fix it in containerd too?

Reply:
Yep, done :)